tests/net/adversary.rs

```diff
@@ -7,14 +7,291 @@
+    ///
+    /// # Panics
+    ///
+    /// Panics if `msg.from` is not a faulty or either `msg.from` or `msg.to` do not exist. Panics
+    /// if `position` is equal to `Before(idx)`, with `idx` being out of bounds.
+    #[inline]
+    pub fn inject_message(&mut self, position: QueuePosition, msg: NetMessage<D>) {
+        // Ensure the node is not faulty.
+        assert!(
+            self.0
+                .get(msg.from.clone())
+                .expect("inject: unknown sender node")
+                .is_faulty(),
+            "Tried to inject message not originating from a faulty node."
+        );
+
+        // Sender must exist.
+        self.0
+            .get(msg.to.clone())
+            .expect("inject: unknown recipient node");
+
+        // Insert into queue. `insert` will panic on out-of-bounds.
+        match position {
+            QueuePosition::Front => self.0.messages.push_front(msg),
+            QueuePosition::Back => self.0.messages.push_back(msg),
+            QueuePosition::Before(idx) => self.0.messages.insert(idx, msg),
+        }
+    }
+
+    /// Swap two messages in the message queue.
+    ///
+    /// # Panics
+    ///
+    /// Panics if either `i` or `j` are out-of-bounds.
+    #[inline]
+    pub fn swap_messages(&mut self, i: usize, j: usize) {
+        self.0.swap_messages(i, j);
+    }
+
+    /// Reorder all messages.
+    ///
+    /// Sorts all message with a comparator function.
+    ///
+    /// Sorting is not cheap, but not prohitibively so, since message queues tend to be small for
+    /// most test cases. See `VirtualNet::sort_messages_by` for notes about sorting efficiency.
+    #[inline]
+    pub fn sort_messages_by<F>(&mut self, f: F)
+    where
+        F: FnMut(&NetMessage<D>, &NetMessage<D>) -> cmp::Ordering,
+    {
+        self.0.sort_messages_by(f)
+    }
+}
+
+// Downgrade-conversion.
+impl<'a, D: DistAlgorithm> From<NetMutHandle<'a, D>> for NetHandle<'a, D> {
+    #[inline]
+    fn from(n: NetMutHandle<D>) -> NetHandle<D> {
+        NetHandle(n.0)
+    }
+}
+
+/// Immutable node handle.
+pub struct NodeHandle<'a, D: 'a>(&'a Node<D>)
+where
+    D: DistAlgorithm;
+
+impl<'a, D: DistAlgorithm> NodeHandle<'a, D>
+where
+    D: DistAlgorithm,
+{
+    /// Construct new immutable node handle.
+    #[inline]
+    fn new(inner: &'a Node<D>) -> Self {
+        NodeHandle(inner)
+    }
+
+    /// Return node ID.
+    #[inline]
+    pub fn id(&self) -> D::NodeId {
+        self.0.id().clone()
+    }
+
+    /// Returns a reference to the faulty node.
+    ///
+    /// Allows access to the wrapped node, if it is faulty.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the node is not faulty.
+    #[inline]
+    pub fn node(&self) -> &'a Node<D> {
+        self.try_node()
+            .expect("could not access inner node of handle, node is not faulty")
+    }
+
+    /// Try to return a reference to the inner faulty node.
+    ///
+    /// If the inner node is faulty, returns a reference to it.
```

Comment:
Duplicate information in docs. I think that `/// If the inner node is faulty, returns a reference to it.` would suffice as the docstring for this method.

Reply:
Fixed.